deepdrivemd/aggregation/stream/aggregator.py

```diff
@@ -0,0 +1,168 @@
+            start = [0]*ndimCM
+            count = shapeCM
+            varCM.SetSelection([start, count])
+            cm = np.zeros(shapeCM, dtype=np.int8)
+            stream.Get(varCM, cm)
+
+            varPositions = io.InquireVariable("positions")
+            shapePositions = varPositions.Shape()
+            ndimPositions = len(shapePositions)
+            start = [0]*ndimPositions
+            count = shapePositions
+            varPositions.SetSelection([start, count])
+            positions = np.zeros(shapePositions, dtype=np.float32)
+            stream.Get(varPositions, positions)
+
+            varVelocities = io.InquireVariable("velocities")
+            shapeVelocities = varVelocities.Shape()
+            ndimVelocities = len(shapeVelocities)
+            start = [0]*ndimVelocities
+            count = shapeVelocities
+            varVelocities.SetSelection([start, count])
+            velocities = np.zeros(shapeVelocities, dtype=np.float32)
+            stream.Get(varVelocities, velocities)
+
+            varMD5 = io.InquireVariable("md5")
+            shapeMD5 = varMD5.Shape()
+            ndimMD5 = len(shapeMD5)
+            start = [0]*ndimMD5
+            count = shapeMD5
+            varMD5.SetSelection([start, count])
+            md5 = np.zeros(shapeMD5, dtype=np.int64)
+            stream.Get(varMD5, md5)
+
+            stream.EndStep()
+            
+            md5 = intarray2hash(md5)
+
+            aggregator_stream.write("md5", md5)
+            aggregator_stream.write("step", step)
+            aggregator_stream.write("rmsd", rmsd)
+            aggregator_stream.write("dir", str(sim_task_id))
+            aggregator_stream.write("positions", positions, list(positions.shape), [0]*len(positions.shape), list(positions.shape))
+            aggregator_stream.write("velocities", velocities, list(velocities.shape), [0]*len(velocities.shape), list(velocities.shape))
+            aggregator_stream.write("contact_map", cm, list(cm.shape), [0]*len(cm.shape), list(cm.shape), end_step=True)
+
+        timer("aggregator_iteration", -1)
+
+if __name__ == "__main__":
+    print(subprocess.getstatusoutput("hostname")[1]); sys.stdout.flush()
+
+    args = parse_args()
+    cfg = StreamAggregation.from_yaml(args.config)
+
+    with Timer("aggregator_find_adios_files"):
+        bpfiles = find_input(cfg)
+    print("bpfiles = ", bpfiles)
+    print("len(bpfiles) = ", len(bpfiles))
+
+    with Timer("aggregator_connect"):
+        connections = connect_to_input(cfg, bpfiles)
+    print("connections = ", connections)
+    print("len(connections) = ", len(connections))
+
+    bpaggregator = str(cfg.output_path/"agg.bp")
+
+    aggregator_stream = adios2.open(name=bpaggregator,
+                                    mode="w", config_file=str(cfg.adios_xml_agg),
+                                    io_in_config_file="AggregatorOutput")
+
+    aggregate(cfg, connections, aggregator_stream)
+
+    aggregator_stream.close()
```

Comment:
Does this line ever get run? The aggregate() function is in a while true loop. Does it not matter that the stream is left open.

Reply:
Indeed in the current program it never gets called due to an infinite loop but if you allocated some resource in constructor, you should release it in destructor. Besides, the usage might change and the loop might become finite.

Reply:
Ah I see, thanks for explaining. Can you please add a comment above that line explaining it?